src/java_bytecode/java_object_factory.cpp

```diff
@@ -494,6 +496,167 @@ class recursion_set_entryt
+/// __CPROVER_assume(tmp_object_factory$1 >= 0);
+/// __CPROVER_assume(tmp_object_factory$1 <= max_nondet_string_length);
+/// char nondet_infinite_array$2[INFINITY()];
+/// nondet_infinite_array$2 = NONDET(char [INFINITY()]);
+/// int return_array$3;
+/// return_array$3 = cprover_associate_array_to_pointer_func
+///   (nondet_infinite_array$2, &nondet_infinite_array$2[0]);
+/// int return_array$4;
+/// return_array$4 = cprover_associate_length_to_array_func
+///   (nondet_infinite_array$2, tmp_object_factory$1);
+/// arg = { .\@java.lang.Object={ .\@class_identifier="java.lang.String",
+///   .\@lock=false }, .length=tmp_object_factory$1,
+///   .data=nondet_infinite_array$2 };
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Unit tests in `unit/java_bytecode/java_object_factory/` ensure
+/// it is the case.
+codet initialize_nondet_string_struct(
+  const exprt &obj,
+  const std::size_t &max_nondet_string_length,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table)
+{
+  PRECONDITION(
+    java_string_library_preprocesst::implements_java_char_sequence(obj.type()));
+
+  const namespacet ns(symbol_table);
+  code_blockt code;
+
+  // `obj` is `*expr`
+  const struct_typet struct_type=to_struct_type(ns.follow(obj.type()));
+  const irep_idt &class_id=struct_type.get_tag();
+
+  // @clsid = String and @lock = false:
+  const symbol_typet jlo_symbol("java::java.lang.Object");
+  const struct_typet jlo_type=to_struct_type(ns.follow(jlo_symbol));
+  struct_exprt jlo_init(jlo_symbol);
+  jlo_init.operands().resize(jlo_type.components().size());
+
+  const std::size_t classid_nb=jlo_type.component_number("@class_identifier");
+  const typet &classid_type=jlo_type.components()[classid_nb].type();
+  jlo_init.operands()[classid_nb]=constant_exprt(class_id, classid_type);
+
+  const std::size_t lock_nb=jlo_type.component_number("@lock");
+  const typet &lock_type=jlo_type.components()[lock_nb].type();
+  jlo_init.operands()[lock_nb]=from_integer(false, lock_type);
+
+  struct_exprt struct_expr(obj.type());
+  struct_expr.copy_to_operands(jlo_init);
+
+  if(struct_type.has_component("length"))
+  {
+    // length_expr = nondet(int);
+    const symbolt length_sym=new_tmp_symbol(symbol_table, loc, java_int_type());
+    const symbol_exprt length_expr=length_sym.symbol_expr();
+    const side_effect_expr_nondett nondet_length(length_expr.type());
+    code.add(code_declt(length_expr));
+    code.add(code_assignt(length_expr, nondet_length));
+
+    // assume (length_expr >= 0);
+    code.add(code_assumet(binary_relation_exprt(
+      length_expr, ID_ge, from_integer(0, java_int_type()))));
+
+    // assume (length_expr <= max_input_length)
+    if(max_nondet_string_length<=max_value(length_expr.type()))
+    {
+      exprt max_length=
+        from_integer(max_nondet_string_length, length_expr.type());
+      code.add(code_assumet(
+        binary_relation_exprt(length_expr, ID_le, max_length)));
```

Comment:
This roughly duplicates the init-length-and-assume-in-bounds code from above, factor out?

Reply:
I'm not sure that is the other code you're talking about

Reply:
Hmm, can't find it again now, never mind.